### DISCUSSION TITLE
Consider using cargo-util-schemas to parse Cargo.toml

### DISCUSSION THREAD

**Discussion by 0xPoe** [rust-lang/crates.io#7811] [Feature Requests]
Consider using [`cargo-util-schemas`](https://github.com/rust-lang/cargo/tree/master/crates/cargo-util-schemas) from the cargo repo. 

We are currently using https://github.com/LukeMathWalker/cargo-manifest to parse the `Cargo.toml`. The Cargo team is planning to release the `cargo-util-schemas`. Using the same library from Cargo directly would help us keep using the same `Cargo.toml` schema from Cargo.

If it's OK, I am willing to help verify that we can use it and migrate to using it.

**Comment by Turbo87** [comment]
Using the same logic as cargo generally makes sense to me. 

cargo-manifest however does have some functionality to e.g. find bins from the filesystem that are not explicitly declared in the manifest. While we are not using this functionality yet, people have been asking us to display binary crates differently from library crates, and to implement that we will need this kind of binary detection. We could also build it on top of `cargo-util-schemas` but I'm a bit weary about reinventing and maintaining another wheel if cargo-manifest and cargo-toml already have done so 😅

**Reply by epage** [reply]
I'm a bit confused by this response as previously the Cargo team [formerly requested](https://github.com/rust-lang/crates.io/pull/6914#issuecomment-1679426764) that crates.io not use these packages and, based on the response, it sounded like there was support for something like `cargo-util-schemas` as a way to meet both teams interests.  This was the motivation for creating `cargo-util-schemas` when we did.

> cargo-manifest however does have some functionality to e.g. find bins from the filesystem that are not explicitly declared in the manifest

I can't remember if I brought this up in the past to the crates.io team but an idea I've been considering is explicitly listing all inferred build-targets in the published `Cargo.toml`.  This would allow crates.io to extract this information with just `cargo-util-schemas` (among other benefits).  This wouldn't allow for backfills but I feel that is less likely to be needed as this is for the UI (as compared to things like `rust-version`) and over time new releases will come out that will include this information.

I've picked this work back up in earnest last week and hope to have results soon (the Cargo blog post has sucked up my last couple of days, which has a section on this).

**Comment by 2ndDerivative** [comment]
I'd give this another try if this hasn't moved forward. It was the original intention of cargo-util-schemas after all
